File: 2015/03/2.py
input = None
try:
    input = open("./2015/03/input.txt", "r")
except:
    print("Error opening input.txt")
    exit(1)


class Coordinate:
    x = 0
    y = 0

    def __init__(self) -> None:
        self.x = 0
        self.y = 0

    def __str__(self) -> str:
        return "({},{})".format(self.x, self.y)


santaPos = Coordinate()
botPos = Coordinate()
pos = botPos
housesWithGifts = set()
housesWithGifts.add(str(santaPos))
for char in input.read():
    if pos == botPos:
        pos = santaPos
    else:
        pos = botPos
    if char == "^":
        pos.y += 1
    elif char == "v":
        pos.y -= 1
    elif char == "<":
        pos.x -= 1
    elif char == ">":
        pos.x += 1
    else:
        print("Error, unknown character '", char, "'", sep="")
        exit(1)
    housesWithGifts.add(str(pos))

print(
    "Number of houses that recieve atleast one gift from santa is {}".format(
        len(housesWithGifts)
    )
)

File: 2015/03/2.js
"use strict";
const fs = require("node:fs");
let contents = fs.readFileSync("./2015/03/input.txt", "utf8").toString();

class Coordinate {
  x;
  y;
  constructor(x, y) {
    this.x = x;
    this.y = y;
  }
  toString() {
    return `(${this.x},${this.y})`;
  }
}

let santaPos = new Coordinate(0, 0);
let botPos = new Coordinate(0, 0);
let pos = botPos;
let housesWithGifts = new Set();
housesWithGifts.add(santaPos.toString());
for (let character of contents) {
  if (pos === botPos) pos = santaPos;
  else pos = botPos;
  switch (character) {
    case "^":
      pos.y++;
      break;
    case "v":
      pos.y--;
      break;
    case "<":
      pos.x--;
      break;
    case ">":
      pos.x++;
      break;
    default:
      console.error(`Unexpected character '${character}' in input file`);
      process.exit(1);
  }
  housesWithGifts.add(pos.toString());
}

console.log(housesWithGifts.size);
